### PR TITLE
MM-17617, MM-17613: Consolidated the Jira API client code

### PR DESCRIPTION
**Pull request by levb**
A little overkill to improve an error message on attaching files... but now:
- Easier to mock out Jira APIs
- More consistent error handling
- With a Client interface, some clarity between server-specific and cloud-specific implementations

cc @mickmister 

Tickets:
- https://mattermost.atlassian.net/browse/MM-17617
- https://mattermost.atlassian.net/browse/MM-17613

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := client.Jira.Do(req, dest)
+	if err != nil {
+		err = userFriendlyJiraError(resp, err)
+	}
+	return err
+}
+
+// RESTPostAttachment uploads an attachment to an issue
```
Comment (levb) [inline server/client.go]: Should probably comment, this is the gist of the bug - the issue/attachments API returns _some_ errors in plain text, and some in JSON. This is a good possibility to contribute back to go-jira, too.

File: server/client.go [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+		return http.StatusOK
+	}
+	coder, ok := err.(StatusCoder)
+	if !ok {
+		return http.StatusInternalServerError
+	}
+	return coder.StatusCode()
+}
+
+func userFriendlyJiraError(resp *jira.Response, err error) error {
```
Comment (cpoile) [inline server/client.go]: This is cool, nice. 👍

File: server/client_cloud.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+			err = errors.New(result["error"])
+		}
+		return nil, RESTError{err, resp.StatusCode}
+	}
+
+	return cimd, nil
+}
+
+// SearchUsersAssignableToIssue finds all users that can be assigned to an issue.
+func (client jiraCloudClient) SearchUsersAssignableToIssue(issueKey, query string, maxResults int) ([]jira.User, error) {
```
Comment (cpoile) [inline server/client_cloud.go]: Why is there a version of this for jiraCloudClient and jiraClient?  (if the only difference is the one param name, maybe pass the param map in?)

File: server/client_server.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		if resp.StatusCode == http.StatusForbidden || resp.StatusCode == http.StatusUnauthorized {
+			err = errors.New("not authorized to create issues")
+		}
+		return nil, RESTError{err, resp.StatusCode}
+	}
+	return cimd, nil
+}
+
+// SearchUsersAssignableToIssue finds all users that can be assigned to an issue.
+func (client JiraClient) SearchUsersAssignableToIssue(issueKey, query string, maxResults int) ([]jira.User, error) {
```
Comment (cpoile) [inline server/client_server.go]: Why is this JiraClient?

File: server/webhook.go [hunk 5]
```diff
@@ -132,31 +132,14 @@ func (wh *webhook) PostNotifications(p *Plugin) ([]*model.Post, int, error) {
...
-
-			resp, _ := jiraClient.Do(req, nil)
-			if resp.StatusCode != http.StatusOK {
-				// recipient does not have permissions to view the issue.
-				continue
-			}
-
+			selfURL = notification.commentSelf
+		}
+		err = client.RESTGet(selfURL, nil, &struct{}{})
```
Comment (cpoile) [inline server/webhook.go]: So much simpler now.

File: server/client_server.go (hunk 4 above)
Comment (levb) [reply server/client_server.go]: Cut&paste error - curiously it just worked because the Server version overrode it, good catch, TY!

**Comment by levb** [comment]
@cpoile please re-review.

File: server/client_server.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+			err = errors.New("not authorized to create issues")
+		}
+		return nil, RESTError{err, resp.StatusCode}
+	}
+	return cimd, nil
+}
+
+// SearchUsersAssignableToIssue finds all users that can be assigned to an issue.
+func (client jiraServerClient) SearchUsersAssignableToIssue(issueKey, query string, maxResults int) ([]jira.User, error) {
+	return SearchUsersAssignableToIssue(client, issueKey, "username", query, maxResults)
```
Comment (cpoile) [inline server/client_server.go]: 👍

**Pull request merged**